Allow directories in exports field of package.json

Node.JS (> 12) allows directories in exports, this works:

```
  "exports": {
    "./": "./dist/"
  },
```

But webpack 5 allows only file aliases

```
  "exports": {
    "./some": "./-/some.js"
  },
```

PS: Demo https://github.com/zerkalica/wp-exports-dir-bug

1. Clone, yarn install
2. cd `lib/b`
3. yarn build

Actual: `Module not found: Error: Can't resolve '@some/lib-a/a' in '.../wp-exports-dir-bug/lib/b/-'`
Expecteed: no errors

To demonstrate, that node works, uncomment `'@some/lib-a/a': require.resolve('@some/lib-a/a')` in lib/b/webpack.config.js
Actual and expected: no errors

--- a/lib/ResolverFactory.js
+++ b/lib/ResolverFactory.js
@@ -377,7 +377,7 @@ exports.createResolver = function (options) {
 				"internal",
 				conditionNames,
 				importsField,
-				"described-relative",
+				"relative",
 				"internal-resolve"
 			)
 		);
@@ -444,7 +444,7 @@ exports.createResolver = function (options) {
 				"resolve-in-package",
 				conditionNames,
 				exportsField,
-				"described-relative"
+				"relative"
 			)
 		);
 	});

--- a/lib/ResolverFactory.js
+++ b/lib/ResolverFactory.js
@@ -377,7 +377,7 @@ exports.createResolver = function (options) {
 				"internal",
 				conditionNames,
 				importsField,
-				"final-file",
+				"described-relative",
 				"internal-resolve"
 			)
 		);
@@ -444,7 +444,7 @@ exports.createResolver = function (options) {
 				"resolve-in-package",
 				conditionNames,
 				exportsField,
-				"final-file"
+				"described-relative"
 			)
 		);
 	});

--- a/test/exportsField.js
+++ b/test/exportsField.js
@@ -1767,10 +1767,12 @@ describe("ExportsFieldPlugin", () => {
 						"        existing directory .../node_modules/exports-field",
 						"          using description file: .../node_modules/exports-field/package.json (relative path: .)",
 						"            using exports field: ./lib/lib2/browser.js",
-						"              .../node_modules/exports-field/lib/lib2/browser.js doesn't exist",
+						"              using description file: .../node_modules/exports-field/package.json (relative path: ./lib/lib2/browser.js)",
+						"                .../node_modules/exports-field/lib/lib2/browser.js doesn't exist",
 						"            using exports field: ./lib/browser.js",
-						"              existing file: .../node_modules/exports-field/lib/browser.js",
-						"                reporting result .../node_modules/exports-field/lib/browser.js"
+						"              using description file: .../node_modules/exports-field/package.json (relative path: ./lib/browser.js)",
+						"                existing file: .../node_modules/exports-field/lib/browser.js",
+						"                  reporting result .../node_modules/exports-field/lib/browser.js"
 					]);
 				done();
 			}

--- a/test/exportsField.js
+++ b/test/exportsField.js
@@ -1316,6 +1316,13 @@ describe("ExportsFieldPlugin", () => {
 	const resolver = ResolverFactory.createResolver({
 		extensions: [".js"],
 		fileSystem: nodeFileSystem,
+		fullySpecified: true,
+		conditionNames: ["webpack"]
+	});
+
+	const commonjsResolver = ResolverFactory.createResolver({
+		extensions: [".js"],
+		fileSystem: nodeFileSystem,
 		conditionNames: ["webpack"]
 	});
 
@@ -1354,7 +1361,7 @@ describe("ExportsFieldPlugin", () => {
 		);
 	});
 
-	it("resolve using exports field, not a browser field #2", done => {
+	it("resolve using exports field and a browser alias field #2", done => {
 		const resolver = ResolverFactory.createResolver({
 			aliasFields: ["browser"],
 			extensions: [".js"],
@@ -1371,7 +1378,7 @@ describe("ExportsFieldPlugin", () => {
 				if (err) return done(err);
 				if (!result) throw new Error("No result");
 				result.should.equal(
-					path.resolve(fixture2, "node_modules/exports-field/lib/main.js")
+					path.resolve(fixture2, "node_modules/exports-field/lib/browser.js")
 				);
 				done();
 			}
@@ -1388,6 +1395,38 @@ describe("ExportsFieldPlugin", () => {
 				if (!err) throw new Error(`expect error, got ${result}`);
 				err.should.be.instanceof(Error);
 				err.message.should.match(/Can't resolve/);
+				done();
+			}
+		);
+	});
+
+	it("throw error if extension not provided", done => {
+		resolver.resolve(
+			{},
+			fixture2,
+			"exports-field/dist/main",
+			{},
+			(err, result) => {
+				if (!err) throw new Error(`expect error, got ${result}`);
+				err.should.be.instanceof(Error);
+				err.message.should.match(/Can't resolve/);
+				done();
+			}
+		);
+	});
+
+	it("should resolve extension without fullySpecified", done => {
+		commonjsResolver.resolve(
+			{},
+			fixture2,
+			"exports-field/dist/main",
+			{},
+			(err, result) => {
+				if (err) return done(err);
+				if (!result) throw new Error("No result");
+				result.should.equal(
+					path.resolve(fixture2, "node_modules/exports-field/lib/lib2/main.js")
+				);
 				done();
 			}
 		);

--- a/test/importsField.js
+++ b/test/importsField.js
@@ -1325,16 +1325,18 @@ describe("ImportsFieldPlugin", () => {
 						"              existing directory .../node_modules/a",
 						"                using description file: .../node_modules/a/package.json (relative path: .)",
 						"                  using exports field: ./lib/lib2/index.js",
-						"                    no extension",
-						"                      .../node_modules/a/lib/lib2/index.js doesn't exist",
-						"                    .js",
-						"                      .../node_modules/a/lib/lib2/index.js.js doesn't exist",
-						"                    as directory",
-						"                      .../node_modules/a/lib/lib2/index.js doesn't exist",
+						"                    using description file: .../node_modules/a/package.json (relative path: ./lib/lib2/index.js)",
+						"                      no extension",
+						"                        .../node_modules/a/lib/lib2/index.js doesn't exist",
+						"                      .js",
+						"                        .../node_modules/a/lib/lib2/index.js.js doesn't exist",
+						"                      as directory",
+						"                        .../node_modules/a/lib/lib2/index.js doesn't exist",
 						"                  using exports field: ./lib/index.js",
-						"                    no extension",
-						"                      existing file: .../node_modules/a/lib/index.js",
-						"                        reporting result .../node_modules/a/lib/index.js"
+						"                    using description file: .../node_modules/a/package.json (relative path: ./lib/index.js)",
+						"                      no extension",
+						"                        existing file: .../node_modules/a/lib/index.js",
+						"                          reporting result .../node_modules/a/lib/index.js"
 					]);
 				done();
 			}

--- a/test/importsField.js
+++ b/test/importsField.js
@@ -1325,10 +1325,16 @@ describe("ImportsFieldPlugin", () => {
 						"              existing directory .../node_modules/a",
 						"                using description file: .../node_modules/a/package.json (relative path: .)",
 						"                  using exports field: ./lib/lib2/index.js",
-						"                    .../node_modules/a/lib/lib2/index.js doesn't exist",
+						"                    no extension",
+						"                      .../node_modules/a/lib/lib2/index.js doesn't exist",
+						"                    .js",
+						"                      .../node_modules/a/lib/lib2/index.js.js doesn't exist",
+						"                    as directory",
+						"                      .../node_modules/a/lib/lib2/index.js doesn't exist",
 						"                  using exports field: ./lib/index.js",
-						"                    existing file: .../node_modules/a/lib/index.js",
-						"                      reporting result .../node_modules/a/lib/index.js"
+						"                    no extension",
+						"                      existing file: .../node_modules/a/lib/index.js",
+						"                        reporting result .../node_modules/a/lib/index.js"
 					]);
 				done();
 			}

